### PR TITLE
CircleCI: Add workflow for building with Grafana Build Pipeline

### PR DESCRIPTION
**Pull request by aknuds1**
**What this PR does / why we need it**:
As part of our work to migrate CI to the [Grafana Build Pipeline](https://github.com/grafana/build-pipeline/) tool, a new CircleCI workflow, "build-pipeline", is introduced. This implements a pipeline for building the master branch, in parallel to the existing "build-master" workflow. Take note that it doesn't replace what's already there.

The new workflow is meant to replace the existing one, but for the time being it's work in progress and needs ironing out of some kinks. Since it's just for testing, it for example fakes any steps involving publishing of artifacts.

You can see an example run of this pipeline, from the grafana/build-pipeline repo, [here](https://circleci.com/workflow-run/684d49a4-cf7b-4f82-b1e4-046a5e4b28c8).

### CODE REVIEW

**Comment by ryantxu** [comment]
On this branch, the build looks like:
image

what does it look like for master/release?

**Comment by aknuds1** [comment]
@ryantxu That CircleCI workflow is [build-branches-and-prs](https://github.com/grafana/grafana/blob/94927428124778e576dc055f747c2735cd7ffcb6/.circleci/config.yml#L1259), which fires for this PR. I.e., it's not the new workflow introduced _by_ this PR, which should only trigger for pushes to master. It's a bit confusing I know :)

File: .circleci/config.yml
Comment (torkelo) [inline .circleci/config.yml]: how do we version this so that a version of the build pipeline version is in sync with grafana branch/version?  Always taking the latest build pipeline seems it will cause issues when we want to build older branches

File: .circleci/config.yml
Comment (torkelo) [inline .circleci/config.yml]: maybe we do not need this if we use google/cloud-sdk image instead for this job

File: .circleci/config.yml
Comment (torkelo) [inline .circleci/config.yml]: what does this do?

File: .circleci/config.yml
Comment (torkelo) [inline .circleci/config.yml]: We do not use appveyor anymore this can be removed

**Review by torkelo — COMMENTED** [review]
This looks really good! nice work @aknuds1 ! 

Really like how it separates building frontend & backend and all the variants.

File: .circleci/config.yml
Comment (torkelo) [inline .circleci/config.yml]: wish there was ways we can define and reuse these lists (it's going to be duplicated for the production release workflow)

Comment (aknuds1) [reply .circleci/config.yml]: @torkelo I think we might be able to re-use the same workflow for release, through branching constructs. I haven't tried it yet, but my impression is that how CircleCI intends for workflows to be used (i.e. one workflow -> many pipelines).

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @torkelo Great! I don't know the history of this even.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @torkelo This centralizes some build logic in the build-pipeline tool: https://github.com/grafana/build-pipeline/blob/master/pkg/subcmd/shellcheck/action.go#L14.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @torkelo Thanks I will try, would be nice to get rid of this step.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @torkelo Agree - my thinking so far is that we version the build pipeline tool and tie Grafana branches to a compatible version.

**Comment by aknuds1** [comment]
> This looks really good! nice work @aknuds1 !
> 
> Really like how it separates building frontend & backend and all the variants.

Thanks @torkelo! I actually tried parallel building all back-end variants in one job, but it OOMs so there's no choice but to splitting into separate jobs.

File: .circleci/config.yml
Comment (xlson) [inline .circleci/config.yml]: I'm wondering if it might be a good idea to incorporate the knowledge of what artifacts are produced in a build step so that grabpl can copy them to the correct location instead of that having to be a separate step in the ci script. Could have an `--ouput oss` arg or something like that.

**Comment by papagian** [comment]
great work!
I have a suggestion: I think that `build-enterprise-backend` slightly differs from `build-oss-backend` therefore they could parameterised by introducing some additional parameters for the different paths. The same for `build-oss-frontend` and `build-enterprise-frontend`

File: .circleci/config.yml
Comment (xlson) [inline .circleci/config.yml]: I don't think this will end up echoing the grafana api key to the logs (I think CircleCI will protect us from that) but it might still be a good idea not to write it like this as it looks like that is what will happen.

File: .circleci/config.yml
Comment (xlson) [inline .circleci/config.yml]: api key again

File: .circleci/config.yml
Comment (xlson) [reply .circleci/config.yml]: Yeah, I think that makes sense.

File: .circleci/config.yml
Comment (xlson) [inline .circleci/config.yml]: I think breaking apart the docker builds to multiple jobs would make a lot of sense too.

**Review by xlson — COMMENTED** [review]
One thing to keep an eye on when building binaries in separate jobs like this is if we can do anything to improve the size (or load time) of our build image. When looking at the last couple of runs of build-all on CircleCI it seems to take about 45-60 seconds to initialize our image.

Are we using the build shellscripts/build.go or does grabpl reimplement these?

**Comment by aknuds1** [comment]
> One thing to keep an eye on when building binaries in separate jobs like this is if we can do anything to improve the size (or load time) of our build image. When looking at the last couple of runs of build-all on CircleCI it seems to take about 45-60 seconds to initialize our image.
> 
> Are we using the build shellscripts/build.go or does grabpl reimplement these?

@xlson grabpl replaces build.go, I don't think it uses it any more. I guess it won't hurt to make the build image smaller and quicker to load.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @xlson Oh yeah, you're right. CircleCI does hide API keys I'm pretty sure, but I shouldn't echo it either. I'll remove it. Thanks!

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @xlson That's probably right. This was hacked together in order to please `persist_to_workspace`, since it doesn't give you particularly good control of where to place files in the workspace, and it's also badly documented.

**Comment by aknuds1** [comment]
> great work!
> I have a suggestion: I think that `build-enterprise-backend` slightly differs from `build-oss-backend` therefore they could parameterised by introducing some additional parameters for the different paths. The same for `build-oss-frontend` and `build-enterprise-frontend`

@papagian Thanks! It's also occurred to me that they OSS and enterprise jobs could be combined into parameterized jobs, but I didn't go that far yet. I can try to do it though, shouldn't be a problem.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @xlson Thanks, will take care of it.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @xlson You mean to break them up based on architecture? I think this workflow only builds AMD64.

Comment (xlson) [reply .circleci/config.yml]: Okay. Yes that is what I meant. Building docker images for all the different architectures takes quite a lot of time

Comment (aknuds1) [reply .circleci/config.yml]: @xlson OK thanks for the clarification.

**Comment by aknuds1** [comment]
@papagian @xlson @torkelo Mind having another look? I tried to incorporate your suggestions, and IMO it should be ready to merge. It doesn't have to be perfect before we merge, to make that clear; the idea is to have it run builds in parallel to the normal master workflow, so we can see how it works out in practice and make adjustments.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @xlson Having taken a closer look, I think Docker images may be built for other architectures as well after all. Do you mind if we take care of dividing the build by architecture in a follow-up PR though? Doing that would take some work, so I'd rather have this PR merged first so we can test the new master pipeline.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: To specify, my thinking is we can come up with a version scheme later, when we're past prototyping.

**Comment by torkelo** [comment]
Think it's fine to merge this if it has no side effects on the current process. But it will cause a slow down in the builds (as more circle ci containers will be devoted to this). So it would be good to have a short timeline between merge and switch over to this build pipeline. After merging this, beyond identifying and fixing issues, is there a big part missing? 

Another option is to merge & test it a few commits and then comment out the trigger until missing pieces are in place

File: .circleci/config.yml
Comment (marefr) [reply .circleci/config.yml]: Why don't use go get? The github token should be able to access that repo. To be able to use it locally would make sense to not have to download tooling from gcp.

File: .circleci/config.yml
Comment (marefr) [reply .circleci/config.yml]: Nit. Would make shellcheck more general by require `"-e", "SC1071", "-e", "SC2162"` to be provided by arguments/config instead of hardcoded in the code.

File: .circleci/config.yml
Comment (marefr) [inline .circleci/config.yml]: Are we publishing to both gcp and grafana.com? Aren't grafana.com storing in gcp?

**Review by marefr — CHANGES_REQUESTED** [review]
I would like to see usage of this new tool in the build-branches and PR's workflow to be able to test run it before merging it to master.

To me seems like you @aknuds1 wants to get it into master to test it out running side by side with existing workflow over some time. @torkelo seems to want to merge to master and replace the existing workflow as fast as possible. 

Given upcoming 6.6 beta and stable, should we wait with merging this until stable is out?

**Comment by aknuds1** [comment]
@marefr I'm going to try a filter for the new workflow that corresponds to PRs/branches, to see if it will be triggered for PRs, so we can test it.

File: .circleci/config.yml
Comment (xlson) [reply .circleci/config.yml]: This can definitely wait until later, no need to do everything at once.

**Comment by xlson** [comment]
Do you want feedback on this PR at this time, and if so, what areas are most important for us to look at?

**Comment by aknuds1** [comment]
> Do you want feedback on this PR at this time, and if so, what areas are most important for us to look at?

Do you see any issues holding the PR back from getting merged?

I still have some feedback from Marcus E that I need to address, and I'd like to do some more verification of the built packages, but apart from that I think it's pretty ready.

**Comment by xlson** [comment]
I'd like us to wait with merging this PR until after the stable release to make it easier to make fixes (if I have the time I would like to add arm docker images for enterprise this week). Other than that there's nothing that I see. When you are done perhaps we should do a group review with you? I think that might be easier as there's so much context needed when looking at the PR.

**Comment by aknuds1** [comment]
Sounds good to me @xlson!

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @marefr I've just copied what we're already doing in the current master pipeline, see `deploy-master` and `deploy-enterprise-master` jobs. I don't really know the rationale behind.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @marefr Is there any need currently for this generalization? My program doesn't differ any from the current Makefile in the respect that it codifies our configuration of shellcheck. My thinking is that once we need different shellcheck configurations, we make it configurable.

Comment (marefr) [reply .circleci/config.yml]: No it's not important for now. Was just a nit/something to think about moving forward.

File: .circleci/config.yml
Comment (marefr) [reply .circleci/config.yml]: My bad. Thought Grafana.com pushed to GCP.

**Review by xlson — APPROVED** [review]
LGTM!

File: package.json
Comment (marefr) [inline package.json]: What does this do? Is it for those that using volta to manage node versions?

**Review by marefr — APPROVED** [review]
This looks great. Such an improvement using parameters :+1: Left some final comments. 

Go linting currently fails, but seems unrelated to these changes. but this could be merged as I see it. 

Merging this to master will run both the old master workflow and the new? Any risk of collisions? Publishing same artifacts etc?

**Comment by aknuds1** [comment]
> This looks great. Such an improvement using parameters  Left some final comments.
> 
> Go linting currently fails, but seems unrelated to these changes. but this could be merged as I see it.
> 
> Merging this to master will run both the old master workflow and the new? Any risk of collisions? Publishing same artifacts etc?

Thanks @marefr! Go linting is unrelated, was fixed by restarting. You're right, when we merge this to the master, it will run in parallel to the old one. It actually doesn't publish any artifacts yet, so that is still handled by the old one. There shouldn't be any risk of collisions since no artifacts are yet published - we already ran this pipeline by letting it be triggered by PR branches, so it shouldn't make any real difference to now merge it into master and let it be triggered by master changes (I already changed the trigger back to being the master branch).

I'm not 100% sure how to make the switch from the old master pipeline to this, but maybe we could change this pipeline (in another PR) to upload artifacts somewhere for testing?

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @marefr It's because of an inconsistency versus how the original master pipeline performs builds - I haven't got around to figuring it out yet. It shouldn't be a blocker from merging, since we need to do more testing and fixing anyway before considering this new pipeline complete.

File: .circleci/config.yml
Comment (aknuds1) [reply .circleci/config.yml]: @marefr OK I'll have a look.

File: package.json
Comment (aknuds1) [reply package.json]: @marefr Good catch, thanks. I actually didn't mean to commit that one. I use it locally, since I use volta instead of nvm, but wasn't meaning to commit it :man_facepalming:

File: .circleci/config.yml
Comment (marefr) [reply .circleci/config.yml]: Okay, please disregard it

File: package.json
Comment (marefr) [reply package.json]: If it adds value to volta users it's probably a good thing, but probably in a separate PR then.

**Comment by marefr** [comment]
>  It actually doesn't publish any artifacts yet

Great. Then I have no more questions/comments :+1: 

Doing the switch we can discuss later

**Pull request merged**
